blocks/matcher.ts

```diff
@@ -25,20 +55,65 @@ const matcherBlock: Block<
   introspect: {
     default: "0",
   },
-  adapt:
-    <TConfig = unknown>({ default: func }: { default: MatchFunc }) =>
-    ($live: TConfig) => {
-      return (ctx: MatchContext) => {
-        const fMatcher = func as unknown as
-          | ((c: TConfig, ctx: MatchContext) => boolean)
-          | MatchFunc;
-        const matcherFuncOrValue = fMatcher($live, ctx);
-        if (typeof matcherFuncOrValue === "function") {
-          return matcherFuncOrValue(ctx);
+  adapt: <TConfig = unknown>(
+    { default: func }: { default: MatchFunc },
+  ) =>
+  (
+    $live: TConfig,
+    httpCtx: HttpContext<
+      { global: unknown; response: { headers: Headers } },
+      unknown
+    >,
+  ) => {
+    const matcherFunc = (ctx: MatchContext) => {
+      const fMatcher = func as unknown as
+        | ((c: TConfig, ctx: MatchContext) => boolean)
+        | MatchFunc;
+      const matcherFuncOrValue = fMatcher($live, ctx);
+      if (typeof matcherFuncOrValue === "function") {
+        return matcherFuncOrValue(ctx);
+      }
+      return matcherFuncOrValue;
+    };
+    const respHeaders = httpCtx.context.state.response.headers;
+    const cacheControl = httpCtx.request.headers.get("Cache-Control");
+    const isNoCache = cacheControl === "no-cache";
+    return (ctx: MatchContext) => {
+      try {
+        let uniqueId = "";
+        // from last to first and stop in the first resolvable
+        // the rational behind is: whenever you enter in a resolvable it means that it can be referenced by other resolvables and this value should not change.
+        for (let i = httpCtx.resolveChain.length - 1; i >= 0; i--) {
+          const chain = httpCtx.resolveChain[i];
+          hasher.hash(typeAsNumber(chain.type));
+          hasher.hash(`${chain.value}`);
+          uniqueId =
+            (`${chain.type}@${chain.value}${uniqueId.length > 0 ? "," : ""}`) +
+            uniqueId;
+          // stop on first resolvable
+          if (chain.type === "resolvable") {
+            break;
+          }
+        }
+        const cookieName = `_dcxf_matchers_${hasher.result()}`;
+        const isMatchFromCookie = isNoCache
+          ? undefined
+          : cookieValue.boolean(getCookies(ctx.request.headers)[cookieName]);
```

Comment:
We should user another header (even Pragma) or signal dev to clear cookies if they want to test matchers.

Seems that No-cache may create some undesirable behavior (two chains with the same resolvable/matcher with different values).

Reply:
> Seems that No-cache may create some undesirable behavior (two chains with the same resolvable/matcher with different values).

would you mind to give me an example of this? the "no-cache" behavior still the same as we used to have on routes selection for audiences, do you have any suggestion instead?

Reply:
We need memoization of the result so we can keep it consistent on the no-cache/no-cookie scenario.

I am also against using existing headers to generate an unexpected behavior, but I think that is ok for now until we create a dev tool to help test A/B and our framework.